metrics-exporter-prometheus/src/lib.rs

```diff
@@ -439,12 +449,31 @@ impl PrometheusBuilder {
         Ok(())
     }
 
+    /// Builds the recorder and returns it.
+    /// This function is only enabled when default features are not set.
+    #[cfg(not(feature = "tokio-exporter"))]
+    pub fn build(self) -> Result<PrometheusRecorder, Error> {
+        let inner = Arc::new(Inner {
+            registry: Registry::new(),
+            distributions: RwLock::new(HashMap::new()),
+            quantiles: self.quantiles.clone(),
+            buckets: self.buckets.clone(),
+            buckets_by_name: self.buckets_by_name,
+            descriptions: RwLock::new(HashMap::new()),
+        });
+
+        let recorder = PrometheusRecorder { inner };
+
+        Ok(recorder)
+    }
+
```

Comment:
The problem here is that you actually need to install `PrometheusRecorder` via something like `metrics::set_boxed_recorder` (see line 429 in your version of the file) otherwise nothing is actually recording the metrics.

I think what I'd like to see overall here is `build`/`install` staying as-is, feature gated as you have them now, but two new variants -- not feature gated because they can exist even if `tokio`/`hyper` are being brought in -- called `build__with_handle`/`install_with_handle`. (I'm not super tied to the naming, but..)

What you would do here is create a new type, something like `PrometheusHandle`, which takes a clone of the inner state -- `Arc<Inner>` -- and instead of putting that new `render` function on `PrometheusRecorder`, you put it on `PrometheusHandle` instead.

Then, you'd emulate `build`/`install` but instead of having them spawn a thread or build a future, just pass back the handle.  This would better match what we have now, giving people the ability to install the recorder as-is or to get it back so they can wrap it in a higher-level construct like `Stack`/`Layer`... albeit with a new-found handle to access the data as they require.

Make sense?

Reply:
Makes sense, I was just looking at set_recorder which I believe takes a reference? So I was thinking we would still have access to the Recorder afterwards. 

I don't mind the build_with_handle/install_with_handle syntax, but just a thought what if instead of having separate functions we just had a helper function on the Recorder that returned the handle?

That way the process would look like this -> Call build, let handle = recorder.get_handle(), then call set_boxed_recorder. 

Just a thought, but I'm open to doing it either way.

Reply:
Let me think about it for a little while.

One thing we definitely need is for `PrometheusHandle` to be its own type.  If we simply alias it over `Arc<Inner>`, we're now exposing other internal types, which is suboptimal.

Reply:
I'll make that switch now

Reply:
Pushed up fixes for type -> struct conversion, and eliminated the now not-needed doc comments on Inner